### PR TITLE
UDF and Expressions support.

### PR DESCRIPTION
**Pull request by mayankshriv**
1. Added TransformOperator and TransformExpressionEvaluator
   functionalities.
2. Added basic built-in functions (add/sub/mult/div).
3. Added Server config for user to specify UDF class.
4. Added unit test for new code.

### CODE REVIEW

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      return;
+    }
+
+    _inited = true;
+    for (String functionName : transformFunctions) {
+      try {
+
+        Class<TransformFunction> functionClass = (Class<TransformFunction>) Class.forName(functionName);
+        TransformFunction transformFunction = functionClass.newInstance();
+        TRANSFORM_FUNCTION_MAP.put(transformFunction.getName().toLowerCase(), transformFunction);
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java]: Is function evaluation case sensitive? For example, is `SELECT myFunc(a)` equivalent to `SELECT MYFUNC(a)`?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/TransformExpressionOperator.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  @Override
+  public Block getNextBlock() {
+    return getNextBlock(new BlockId(_nextBlockCallCounter++));
+  }
+
+  @Override
+  public Block getNextBlock(BlockId blockId) {
+    // Only one block supported, currently.
+    if (blockId.getId() > 0) {
+      return null;
```
Comment (jfim) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/TransformExpressionOperator.java]: Should an exception be thrown if we reach an unsupported branch?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DefaultExpressionEvaluator.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+      String column = expressionTree.getColumn();
+
+      if (column != null) {
+        if (isNumeric(column)) {
+          return new DoubleArrayTransformResult(_dataBlockCache.getDoubleValueArrayForColumn(column));
+        } else {
+          throw new RuntimeException("Non numeric columns not supported in transforms yet");
+        }
+      } else {
+        throw new RuntimeException("Literals not supported in transforms yet");
```
Comment (jfim) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DefaultExpressionEvaluator.java]: Are we planning to support constants in another PR? (eg. `SELECT foo + 1`)

Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DefaultExpressionEvaluator.java]: Yes, constants will be supported in another PR.

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java (hunk 1 above)
Comment (mayankshriv) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java]: It is case insensitive.

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+
+/**
+ * Class for representing expression trees for transforms.
+ * <p> - A TransformExpressionTree node has either transform function or a column name, or a literal.</p>
+ * <p> - Leaf nodes either have column name or literal, whereas non-leaf nodes have transform function.</p>
+ * <p> - Transform function in non-leaf nodes is applied to its children nodes. </p>
+ */
+public class TransformExpressionTree {
+  private final int _id;
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java]: Any reason why we need this id?

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+   * Constructor for the class
+   * @param id Unique id for the instance
+   * @param column Column on which to apply the transform.
+   * @param literal A constant value, to be used as an argument for the transform function
+   * @param transformName Name of transform function to apply
+   * @param children Children nodes for the current node
+   */
+  public TransformExpressionTree(int id, String column, String literal, String transformName,
+      List<TransformExpressionTree> children) {
+    super();
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java]: super()? The constructor for Object?

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+  /**
+   * Returns an instance of TransformFunction for the given name. Returns null if
+   * function name was not found, is null, or could not instantiate the class.
+   *
+   * @param functionName Transform function name
+   * @return TransformFunction for the given functionName
+   */
+  public static TransformFunction get(String functionName) {
+    if (functionName == null) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java]: Maybe a check for not null is better here? (In my opinion, should not pass in null as an argument in get())

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  /**
+   * This method builds the transform factory containing functions
+   * specified in the server configuration. Throws {@link RuntimeException} if it has
+   * already been initialized. Method is synchronized (as opposed to concurrent), as it is
+   * expected to be called only once, during start-up.
+   *
+   * @param transformFunctions Array of transform function names
+   */
+  @SuppressWarnings("unchecked")
+  public static synchronized void init(@Nonnull String[] transformFunctions) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java]: If init() is synchronized, then get() should be synchronized as well? Or at least put "_inited = true" as the last statement in this method?
Also, it looks a little weird to make the init() a static method. Since we need to pass in the transformFunctions, why not just hardcode a map for this class?

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  /**
+   * Given the root node for AST tree, builds the expression tree, and returns
+   * the root node of the expression tree.
+   *
+   * @param rootNode Root Node of AST tree
+   * @param id Node ID
+   * @return Root node of the Expression tree
+   */
+  public static TransformExpressionTree buildTree(AstNode rootNode, int id) {
+    TransformExpressionTree expressionTree;
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java]: (OCD) No need for this variable.

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+      } else {
+        throw new Pql2CompilationException(
+            "Leaf of expression tree can only be identifier or literal, found " + rootNode.getClass().getName());
+      }
+    } else if (rootNode instanceof FunctionCallAstNode) {
+      List<TransformExpressionTree> children = new ArrayList<>();
+      expressionTree =
+          new TransformExpressionTree(id, null, null, ((FunctionCallAstNode) rootNode).getName(), children);
+      for (AstNode child : rootNode.getChildren()) {
+        children.add(buildTree(child, ++id));
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java]: This logic cannot support nested functions, there will be nodes with same id.

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+   */
+  public static TransformFunction get(String functionName) {
+    if (functionName == null) {
+      return null;
+    }
+
+    Preconditions.checkState(_inited, "TransformFunctionFactory not initialized.");
+    String key = functionName.toLowerCase();
+
+    if (!TRANSFORM_FUNCTION_MAP.containsKey(key)) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java]: "return TRANSFORM_FUNCTION_MAP.get(key);" is enough.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DefaultExpressionEvaluator.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  /**
+   * Helper (recursive) method that walks the expression tree bottom up evaluating
+   * transforms at each level.
+   *
+   * @param expressionTree Expression tree to evaluate
+   * @return Result of the expression transform
+   */
+  private TransformResult evaluateExpression(TransformExpressionTree expressionTree) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DefaultExpressionEvaluator.java]: No need to pass in expressionTree as it is already a member variable.

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunction.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+   *
+   * All input double[] are assumed to be of same length, and checks for the
+   * same can be skipped for performance reasons.
+   *
+   *
+   * @param length Length of doc ids to process
+   * @param input Array of input values
+   * @return TransformResult object containing transformed values.
+   */
+  TransformResult transform(int length, Object... input);
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunction.java]: Do we need the "length"? Is "length" the length of the "input" array? (The javaDoc for this method is a little confusing)

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DefaultExpressionEvaluator.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    TransformFunction function = expressionTree.getTransform();
+
+    if (function != null) {
+      List<TransformExpressionTree> children = expressionTree.getChildren();
+      Object[] transformArgs = new Object[children.size()];
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DefaultExpressionEvaluator.java]: Cache children.size()

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/TransformExpressionOperator.java [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+   * Constructor for the class
+   *
+   * @param indexSegment Segment to process
+   * @param projectionOperator Projection operator
+   * @param expressionTree Expression tree to evaluate
+   */
+  public TransformExpressionOperator(IndexSegment indexSegment, MProjectionOperator projectionOperator,
+      TransformExpressionTree expressionTree) {
+
+    Preconditions.checkArgument((indexSegment != null));
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/TransformExpressionOperator.java]: checkNotNull()

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/TransformExpressionOperator.java [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    int numDocsScanned = 0;
+    final long startTimeMillis = System.currentTimeMillis();
+
+    _expressionEvaluator.init();
+    while (_projectionOperator.nextBlock() != null) {
+      ProjectionBlock currentBlock = _projectionOperator.getCurrentBlock();
+      Block docIdSetBlock = currentBlock.getDocIdSetBlock();
+      numDocsScanned = processBlock(currentBlock, docIdSetBlock, numDocsScanned);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/TransformExpressionOperator.java]: No need to pass in docIdSetBlock because whether it can be used depends on the class of currentBlock.
Also, no need to pass in numDocsScanned, just return the docsScanned for the currentBlock, and add it to the numDocsScanned.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/TransformExpressionOperator.java [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    _expressionEvaluator.finish();
+
+    IntermediateResultsBlock result = new IntermediateResultsBlock(_expressionEvaluator.getResult());
+    result.setTotalRawDocs(_indexSegment.getSegmentMetadata().getTotalRawDocs());
+    result.setTimeUsedMs(System.currentTimeMillis() - startTimeMillis);
+
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/TransformExpressionOperator.java]: Set numDocsScanned?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/TransformExpressionOperator.java [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+    if (block instanceof DocIdSetBlock) {
+      DocIdSetBlock docIdSetBlock = (DocIdSetBlock) block;
+      int length = docIdSetBlock.getSearchableLength();
+
+      _expressionEvaluator.evaluate(docIdSetBlock.getDocIdSet(), length);
+      numDocsScanned += length;
+
+    } else {
+
+      BlockDocIdSet blockDocIdSet = projectionBlock.getBlockDocIdSet();
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/TransformExpressionOperator.java]: For ProjectionBlock, we have both getDocIdSetBlock() and getBlockDocIdSet()? LOL

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DefaultExpressionEvaluator.java (hunk 13 above)
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DefaultExpressionEvaluator.java]: Done

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java (hunk 10 above)
Comment (mayankshriv) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java]: Done

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java (hunk 6 above)
Comment (mayankshriv) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java]: Fixed.

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java (hunk 9 above)
Comment (mayankshriv) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java]: Removed id.

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java (hunk 4 above)
Comment (mayankshriv) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java]: Removed it.

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java (hunk 5 above)
Comment (mayankshriv) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformExpressionTree.java]: Removed.

File: pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java (hunk 7 above)
Comment (mayankshriv) [inline pinot-common/src/main/java/com/linkedin/pinot/common/request/transform/TransformFunctionFactory.java]: Moved _inited = true as the last statement.

Not sure if I follow your second comment, the transform functions can be configured via server config, so those values cannot be hardcoded in the factory.

**Comment by Jackie-Jiang** [comment]
Why the output Block of TransformOperator is IntermediateResultsBlock? It should be a ProjectionBlock like Block which can be passed to Aggregation/Selection Operator for further calculation right? The getNextBlock() in TransforemOperator should be called multiple times to get a trunk of values each time (similar to ProjectionOperator), and it should be in the middle of ProjectionOperator and Aggregation/Selection Operator. How do you plan to hook the TransformOperator with other Operators?

**Comment by mayankshriv** [comment]
@Jackie-Jiang The output of transform operator will be TransformBlock, which will most likely be extending ProjectionBlock. That will be done in the upcoming PR which will also contain the transform handling during planning phase. Unless you would absolutely want that change within this PR as well, I prefer to keep the planning phase related changes in the upcoming PR which will also address this comment.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/result/DoubleArrayTransformResult.java [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+  private final double[] _values;
+
+  /**
+   * Constructor for the class.
+   *
+   * @param values Result values to store.
+   */
+  public DoubleArrayTransformResult(double[] values) {
+    this._values = values;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/result/DoubleArrayTransformResult.java]: Remove "this."

File: pinot-core/src/test/java/com/linkedin/pinot/query/transform/AdditionTransformTest.java [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+import com.linkedin.pinot.core.operator.transform.function.AdditionTransform;
+import java.util.Random;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+
+/**
+ * Test of {@link AdditionTransform} class.
+ */
+public class AdditionTransformTest {
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/com/linkedin/pinot/query/transform/AdditionTransformTest.java]: Better to wrap these 4 tests into one? Up to you.

**Comment by mayankshriv** [comment]
Will Handle remaining comments in upcoming PR.
